Reland "[LTO] Run Argument Promotion before IPSCCP"

hazzlim opened the pull request:
Run ArgumentPromotion before IPSCCP in the LTO pipeline, to expose more constants to be propagated. We also run PostOrderFunctionAttrs to improve the information available to ArgumentPromotion's alias analysis, and SROA to clean up allocas.

Relands #111163.

hazzlim committed:
Reland "[LTO] Run Argument Promotion before IPSCCP" (#111839)

Run ArgumentPromotion before IPSCCP in the LTO pipeline, to expose more
constants to be propagated. We also run PostOrderFunctionAttrs to
improve the information available to ArgumentPromotion's alias analysis,
and SROA to clean up allocas.

hazzlim commented:
Polite ping on this :)

nikic reviewed (commented):
Given how low the compile-time impact is, I'm inclined to accept this.

I see that there is a PostOrderFunctionAttrsPass run shortly after these passes -- would it be feasible to move it earlier rather than running twice?

In any case, this change needs a PhaseOrdering test.

hazzlim committed:
Precommit test for "[LTO] Run Argument Promotion before IPSCCP"

hazzlim committed:
Move POFA earlier instead of running twice (For >01)

hazzlim commented:
> In any case, this change needs a PhaseOrdering test.

Thanks for pointing this out - I've added a PhaseOrdering test.

> I see that there is a PostOrderFunctionAttrsPass run shortly after these passes -- would it be feasible to move it earlier rather than running twice?

Good point - I've moved the `PostOrderFunctionAttrsPass` from after IPSCCP to before, and this doesn't seem to break any PhaseOrdering tests or cause any regressions in SPEC2017.

nikic commented on llvm/lib/Passes/PassBuilderPipelines.cpp:
```diff
@@ -1840,9 +1849,12 @@ PassBuilder::buildLTODefaultPipeline(OptimizationLevel Level,
-  // Now deduce any function attributes based in the current code.
-  MPM.addPass(
-      createModuleToPostOrderCGSCCPassAdaptor(PostOrderFunctionAttrsPass()));
+  // For higher optimization levels this Pass has just run, so don't repeat it.
```
I don't think this is needed, LTO at `-O1` is not really something anybody cares about.

hazzlim replied:
Ah ok - I have updated to remove running `PostOrderFunctionAttrsPass()` here at `-O1`.

hazzlim committed:
Don't bother running POFA Pass at -01

nikic commented on llvm/lib/Passes/PassBuilderPipelines.cpp:
```diff
@@ -1828,6 +1828,15 @@ PassBuilder::buildLTODefaultPipeline(OptimizationLevel Level,
+    MPM.addPass(
+        createModuleToPostOrderCGSCCPassAdaptor(ArgumentPromotionPass()));
+    MPM.addPass(
+        createModuleToFunctionPassAdaptor(SROAPass(SROAOptions::ModifyCFG)));
```
Sorry for not noting this before: Can we make all of these part of a single CGSCC pass manager? I don't think it makes sense to do a separate SCC iteration for FunctionAttrs and ArgPromo.

hazzlim replied:
No worries - I have updated to use a single CGSCCPassManager for FunctionAttrs and ArgPromo.

nikic replied:
You should be able to add SROA to it as well, with a CGSCC to Function pass adaptor. (I think this may also give better results? Otherwise we'll only run SROA after *all* arg promotion, rather than interleaved.)

hazzlim replied:
Ah I missed that - I have updated to add SROA to the CGSCCPassManager.

hazzlim committed:
Use a single CGSCCPassManager

hazzlim committed:
Also run SROA in the CGSCCPassManager

nikic reviewed (approved):
LGTM

hazzlim commented:
Thank you for reviewing :)

thurstond commented:
An msan buildbot (https://lab.llvm.org/buildbot/#/builders/164/builds/4320) started breaking around the time of this change; this change or https://github.com/llvm/llvm-project/pull/114964 are the only non-revert non-MLIR changes. Could you please take a look?

```
stage2/msan check

FAIL: Clang :: CXX/temp/temp.decls/temp.spec.partial/temp.spec.partial.member/p2.cpp (2331 of 86924)
******************** TEST 'Clang :: CXX/temp/temp.decls/temp.spec.partial/temp.spec.partial.member/p2.cpp' FAILED ********************
Exit Code: 1
Command Output (stderr):
--
RUN: at line 1: /home/b/sanitizer-x86_64-linux-bootstrap-msan/build/llvm_build_msan/bin/clang -cc1 -internal-isystem /home/b/sanitizer-x86_64-linux-bootstrap-msan/build/llvm_build_msan/lib/clang/20/include -nostdsysteminc -std=c++20 -fsyntax-only -verify /home/b/sanitizer-x86_64-linux-bootstrap-msan/build/llvm-project/clang/test/CXX/temp/temp.decls/temp.spec.partial/temp.spec.partial.member/p2.cpp
+ /home/b/sanitizer-x86_64-linux-bootstrap-msan/build/llvm_build_msan/bin/clang -cc1 -internal-isystem /home/b/sanitizer-x86_64-linux-bootstrap-msan/build/llvm_build_msan/lib/clang/20/include -nostdsysteminc -std=c++20 -fsyntax-only -verify /home/b/sanitizer-x86_64-linux-bootstrap-msan/build/llvm-project/clang/test/CXX/temp/temp.decls/temp.spec.partial/temp.spec.partial.member/p2.cpp
error: 'expected-error' diagnostics seen but not expected: 
  File /home/b/sanitizer-x86_64-linux-bootstrap-msan/build/llvm-project/clang/test/CXX/temp/temp.decls/temp.spec.partial/temp.spec.partial.member/p2.cpp Line 73: static assertion failed due to requirement 'A<int>::B<int &>::y == 4'
  File /home/b/sanitizer-x86_64-linux-bootstrap-msan/build/llvm-project/clang/test/CXX/temp/temp.decls/temp.spec.partial/temp.spec.partial.member/p2.cpp Line 77: static assertion failed due to requirement 'A<int>::x<int &> == 4'
error: 'expected-note' diagnostics seen but not expected: 
  File /home/b/sanitizer-x86_64-linux-bootstrap-msan/build/llvm-project/clang/test/CXX/temp/temp.decls/temp.spec.partial/temp.spec.partial.member/p2.cpp Line 73: expression evaluates to '3 == 4'
  File /home/b/sanitizer-x86_64-linux-bootstrap-msan/build/llvm-project/clang/test/CXX/temp/temp.decls/temp.spec.partial/temp.spec.partial.member/p2.cpp Line 77: expression evaluates to '3 == 4'
```

thurstond commented:
My bad, please ignore my above message. The failure was due to a revert (https://github.com/llvm/llvm-project/commit/f548d39c3c751446d124c08769080214680d53ba).

hazzlim commented:
> My bad, please ignore my above message. The failure was due to a revert ([f548d39](https://github.com/llvm/llvm-project/commit/f548d39c3c751446d124c08769080214680d53ba)).



> My bad, please ignore my above message. The failure was due to a revert ([f548d39](https://github.com/llvm/llvm-project/commit/f548d39c3c751446d124c08769080214680d53ba)).

Ah ok, thanks for the update!
